### PR TITLE
Remap individual ids after simplification

### PR DESCRIPTION
**Pull request by benjeffery**
@jeromekelleher  Just trying to get my head round this by adding it to the python simplify first. Is it this simple if we aren't worrying about consistency with the genetic info?

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Er, yes, actually, I think that basically does it. There'll be some subtleties, but that should be the fundamental change we need, since we're already remapping the individual IDs.

What we really want here to test this stuff out is a way of simulating individuals in the simple forward simulator we have. I think we should be able to add an option ``record_individuals=False`` option and add an individual [here](https://github.com/tskit-dev/tskit/blob/main/python/tests/test_wright_fisher.py#L154) and it should all work itself out, ish.

That way we can get some decent sized test data without having to hand-craft tree sequences.

**Comment by benjeffery** [comment]
> What we really want here to test this stuff out is a way of simulating individuals in the simple forward simulator we have. I think we should be able to add an option `record_individuals=False` option and add an individual [here](https://github.com/tskit-dev/tskit/blob/main/python/tests/test_wright_fisher.py#L154) and it should all work itself out, ish.

I've had a stab at that in https://github.com/tskit-dev/tskit/pull/1186/commits/0ad085bd0327859529ca70359c7284468ddfd98e adding the same level of testing that migrations have - which seems to just be counting the rows.

**Comment by jeromekelleher** [comment]
> I've had a stab at that in 0ad085b adding the same level of testing that migrations have - which seems to just be counting the rows.

Nice! I'd need to take the code for a spin to be sure I know what's happening, but can't right now (should have started this presentation yesterday...)

**Comment by benjeffery** [comment]
> > I've had a stab at that in [0ad085b](https://github.com/tskit-dev/tskit/commit/0ad085bd0327859529ca70359c7284468ddfd98e) adding the same level of testing that migrations have - which seems to just be counting the rows.
> 
> Nice! I'd need to take the code for a spin to be sure I know what's happening, but can't right now (should have started this presentation yesterday...)

No worries - I'm working on the c side now.

**Comment by jeromekelleher** [comment]
Working on it now - will tack in a commit with some updates to the Python code.

**Comment by jeromekelleher** [comment]
@benjeffery I pushed a commit with some updates. There was a slight logic bug in the individual ID remapping and I fixed up the tests a bit. The test refactoring could probably be done better, but it's a big improvement over what it was anyway.

**Comment by jeromekelleher** [comment]
So the tests to be done are now to see if the identity relationships between nodes and individuals are retained through simplification. The easiest way to do that I think is to bung in some metadata into both, simplify, and then see if every node's individual has the same metadata as the node. Then, for those parents that happen to survive simplification (which probably won't be many, most will be associated with unary nodes), the parent relationships should be the same.

For ``filter_individuals``=False, we should just keep the whole pedigree in place.

**Comment by jeromekelleher** [comment]
@hyanwong, this addition to the simulator will make your life easier in #1154.

**Comment by petrelharp** [comment]
This is going to be so much easier to test now that we have metadata!

**Comment by benjeffery** [comment]
I've added some tests using the forward sim.

**Comment by benjeffery** [comment]
Ok, there is one last thing to decide here - in #1192 @jeromekelleher suggested simplify should cope with unordered individuals. The current re-referencing code in this PR only works with sorted individuals as it works in the same pass building the new table. It's no biggie to add another pass over the individual table though.

**Comment by jeromekelleher** [comment]
@petrelharp, your call here. A second linear pass through the individuals is no biggie IMO, and probably worth not enforcing a full topological sort on the input data for.

**Comment by petrelharp** [comment]
Sounds good - let's do the second pass. SLiM currently might have individuals before their parents (because the 'remembered' ones come up top of the table, so if a child is remembered and its parent is not...), and we could change that, but it might actually be a can of worms (because (a) if a child gets remembered before a parent, they'd be out of order; or (b) when we output, we stick the current generation on to the individual table, with the remembered indiviiduals up top; to enforce the order in these two situations we'd have to do the topological sort...).

**Comment by benjeffery** [comment]
@jeromekelleher I've added tests with shuffled individuals (that triggered `tsk_bug_assert`), then modified the C to cope. Should be ready to go after a squash.

File: python/tests/test_tables.py [hunk 1]
```diff
@@ -2361,6 +2362,134 @@ def test_samples_interface(self):
...
+        # Check we have a mixed up order
+        with pytest.raises(
+            tskit.LibraryError,
+            match="Individuals must be provided in an order where"
+            " children are after their parent individuals",
+        ):
+            tables.tree_sequence()
+
+        tables.simplify()
+        metadata = [
```
Comment (jeromekelleher) [inline python/tests/test_tables.py:2453]: Might be easier make a tree sequence out of it, and let machinery do the metadata processing for you?

**Review by jeromekelleher — COMMENTED** [review]
LGTM, some minor suggestions to reduce the amount of code.

File: python/tests/test_tables.py [hunk 2]
```diff
@@ -2361,6 +2362,134 @@ def test_samples_interface(self):
...
+                if parent != tskit.NULL:
+                    assert (
+                        ts.individual(parent).metadata["original_id"] == original_parent
+                    )
+        assert set(tables.individuals.parents) != {-1}
+
+    def test_shuffled_individual_parent_mapping(self, wf_sim_with_individual_metadata):
+        tables = wf_sim_with_individual_metadata.copy()
+        individuals = tables.individuals
+
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tests/test_tables.py:2410]: Didn't know about that method! Doh!

File: python/tests/test_tables.py (hunk 1 above)
Comment (benjeffery) [reply python/tests/test_tables.py:2453]: We can't as the individuals being shuffled violates the requirements for a tree sequence. Hence the `raises` check above.

Comment (jeromekelleher) [reply python/tests/test_tables.py:2453]: Doesn't ``simplify`` sort this out, or is that a follow on?

Comment (benjeffery) [reply python/tests/test_tables.py:2453]: No, simplify doesn't sort the individuals. You'd have to do a `sort_tables` after, but then you're not just testing simplify.

Comment (jeromekelleher) [reply python/tests/test_tables.py:2453]: Fair enough :+1:

**Comment by benjeffery** [comment]
@jeromekelleher Comments addressed, thanks for pointing out the existing shuffle method, I should have had a look for one!

**Review by jeromekelleher — APPROVED** [review]
LGTM!

**Pull request merged**